src/main/scala/com/cognite/spark/datasource/TimeSeriesRelation.scala

```diff
@@ -33,301 +30,98 @@ case class TimeSeriesItem(
-                updateTimeSeriesItems,
-                updateTimeSeriesUrl,
-                conflict.error)
-            case Left(_) => IO.raiseError(onError(updateTimeSeriesUrl, response))
-          }
-      }
+  override def getFromRowAndCreate(rows: Seq[Row]): IO[Unit] = {
+    val timeSeriesSeq = rows.map { r =>
+      val timeSeries = fromRow[TimeSeries](r)
+      timeSeries.copy(metadata = filterMetadata(timeSeries.metadata))
     }
-    (updateItems, postItems).parMapN((_, _) => ())
+
+    client.timeSeries
+      .createFromRead(timeSeriesSeq)
+      .handleErrorWith {
+        case e: com.cognite.sdk.scala.common.CdpApiException =>
+          if (e.code == 409) {
+            val existingExternalIds =
+              e.duplicated.get.map(j => j("externalId").get.asString.get)
+            resolveConflict(existingExternalIds, timeSeriesSeq)
+          } else { IO.raiseError(e) }
+      } *> IO.unit
   }
 
   def resolveConflict(
-      timeSeriesItems: Seq[TimeSeriesItem],
-      timeSeriesUrl: Uri,
-      updateTimeSeriesItems: Seq[UpdateTimeSeriesItem],
-      updateTimeSeriesUrl: Uri,
-      duplicateIds: TimeSeriesConflict): IO[Unit] = {
-    implicit val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
+      existingExternalIds: Seq[String],
+      timeSeriesSeq: Seq[TimeSeries]): IO[Unit] = {
+    import CdpConnector.cs
+    val (timeSeriesToUpdate, timeSeriesToCreate) = timeSeriesSeq.partition(
+      p => existingExternalIds.contains(p.externalId.get)
+    )
 
-    val notFound = duplicateIds.notFound
-    val timeSeriesToUpdate = updateTimeSeriesItems.filter(p => !notFound.contains(p.id))
-    val putItems = if (timeSeriesToUpdate.isEmpty) {
-      IO.unit
-    } else {
-      post(
-        config,
-        updateTimeSeriesUrl,
-        timeSeriesToUpdate
-      )
-    }
+    val idMap = client.timeSeries
+      .retrieveByExternalIds(existingExternalIds)
+      .map(_.map(ts => ts.externalId -> ts.id).toMap)
+      .unsafeRunSync()
 
-    val timeSeriesToCreate = timeSeriesItems.filter(p => p.id.exists(notFound.contains(_)))
-    val postTimeSeriesToCreate = timeSeriesToCreate.map(t => PostTimeSeriesItem(t))
-    val postItems = if (timeSeriesToCreate.isEmpty) {
-      IO.unit
-    } else {
-      post(
-        config,
-        timeSeriesUrl,
-        postTimeSeriesToCreate
-      ).flatTap { _ =>
-        IO {
-          if (config.collectMetrics) {
-            timeSeriesCreated.inc(postTimeSeriesToCreate.length)
-          }
-        }
+    val create =
+      if (timeSeriesToCreate.isEmpty) { IO.unit } else {
+        client.timeSeries.createFromRead(timeSeriesToCreate)
+      }
+    val update =
+      if (timeSeriesToUpdate.isEmpty) { IO.unit } else {
+        client.timeSeries.updateFromRead(timeSeriesToUpdate.map(ts =>
+          ts.copy(id = idMap(ts.externalId))))
       }
-    }
 
-    (putItems, postItems).parMapN((_, _) => ())
+    (create, update).parMapN((_, _) => ())
```

Comment:
Do we still log something useful if something fails?

Reply:
This will log any exception returned from the SDK upon failure.